File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_CDKEY.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Linq;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_CDKEY : Message
    {
        public const string KEYOWNER_TOOMANYSPAWNS = "TOO MANY SPAWNS";
        public const string KEYOWNER_NOSPAWNING = "NO SPAWNING";

        public enum Statuses : UInt32
        {
            Success = 1,
            InvalidKey = 2,
            BadProduct = 3,
            Banned = 4,
            InUse = 5,
        }

        public SID_CDKEY()
        {
            Id = (byte)MessageIds.SID_CDKEY;
            Buffer = new byte[0];
        }

        public SID_CDKEY(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_CDKEY;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                        if (Buffer.Length < 6)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be at least 6 bytes");
                        /**
                         * (UINT32) Spawn Key (1 is TRUE, 0 is FALSE)
                         * (STRING) Game Key
                         * (STRING) Key owner name *
                         */

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        context.Client.GameState.SpawnKey = r.ReadUInt32() == 1;
                        context.Client.GameState.GameKeys.Append(new GameKey(r.ReadString()));
                        context.Client.GameState.KeyOwner = r.ReadByteString();

                        return new SID_CDKEY().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT32) Result
                         * (STRING) Key Owner
                         */

                        Buffer = new byte[5];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((UInt32)Statuses.Success);
                        w.Write((byte)0);

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_CDKEY.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Linq

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_CDKEY
        Inherits Message

        Public Const KEYOWNER_TOOMANYSPAWNS As String = "TOO MANY SPAWNS"
        Public Const KEYOWNER_NOSPAWNING As String = "NO SPAWNING"

        Public Enum Statuses As UInt32
            Success = 1
            InvalidKey = 2
            BadProduct = 3
            Banned = 4
            InUse = 5
        End Enum

        Public Sub New()
            Id = CByte(MessageIds.SID_CDKEY)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_CDKEY)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 6 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be at least 6 bytes")

                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            context.Client.GameState.SpawnKey = r.ReadUInt32() = 1
                            Dim unused = context.Client.GameState.GameKeys.Append(New GameKey(r.ReadString()))
                            context.Client.GameState.KeyOwner = r.ReadByteString()
                        End Using
                    End Using

                    Return New SID_CDKEY().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient))
                Case MessageDirection.ServerToClient
                    Buffer = New Byte(4) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CUInt(Statuses.Success))
                            w.Write(CByte(0))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
